### PR TITLE
persist: separate writing updates to blob store from appending to shard metadata

### LINKED ISSUE #12468: persist: adjust write API so s3 write and state machine application can be separately retried at the persist user level
An `append()` (or `compare_and_append()`) call internally does two separate steps:

1. write the batch of updates to a `Blob` implementation (S3, in production)
2. add a pointer to the new batch in the metadata (aka state machine), which we keep in a `Consensus` implementation

We don't currently retry the second step, so a failure can void the complete append attempt, which has to be retried as a whole. Instead, we should separate the two parts at the API level to allow writing the batch of updates separately (which we can already retry internally in the client) and then retry "appending" it into the shard metadata.

Background context: we don't retry the second step, because we cannot ensure correctness for some usage patterns.

### PR DESCRIPTION
**Pull request by aljoscha**
An `append()` (or `compare_and_append()`) call internally does two
separate steps:

1. write the batch of updates to a Blob implementation (S3, in
   production)
2. add a pointer to the new batch in the metadata (aka state machine),
   which we keep in a Consensus implementation

We don't currently retry the second step, so a failure can void the
complete append attempt, which has to be retried as a whole.

Now, we allow separating the two parts at the API level to allow writing
the batch of updates separately (which we can already retry internally
in the client) and then retry "appending" it into the shard metadata.

Background context: we don't retry the second step, because we cannot
ensure correctness for some usage patterns.

Closes #12468

### Tips for reviewer

I didn't add tests yet but I see two options for that:

1. Rewrite all existing tests to use the new API, add a sanity-check to make sure the old-style API works as expected
2. Leave all tests as they are, add sanity-check tests to make sure the new-style API works as expected

Option 2. would be less disruptive but feels wrong because the old-style APIs are really sugar on top of the new APIs, so it seems a sanity-check test is enough for them.

Option 1. is less disruptive in terms of number of lines of code we'd have to change. 

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### CODE REVIEW

**Comment by aljoscha** [comment]
Also, the `_core` methods currently consume the `Batch`, meaning you can't easily retry. I'll have to address that.

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:419]: I'm curious if users will always have the `upper` at the very beginning of the call. imagine a world where a client wants to buffer stuff while the last call to persist is still ongoing. if we only required the lower when the `BatchBuilder` starts, it would allow the user to start writing stuff to s3 while the last compare_and_append call was ongoing. then when the compare_and_append call finishes, they stop adding stuff to the BatchBuilder and immediately hand it off to compare_and_append and start up the next BatchBuffer

to implement it, the BatchBuilder would be initialized with only a `batch_lower`. it internally keeps track of a `batch_upper` which is initially set to `batch_lower`. as updates are added, we require they are `> batch_lower` and advance `batch_upper` so it's always greater than the update ts. then the compare_and_append call would validate `expected_upper` and `upper` in terms of `batch_lower` and `batch_upper`

I may not be explaining this well, so happy to chat if it doesn't make sense. also possible this should be a followup PR, but i'd like to explore if it's possible/desirable before we make the API changes in this one :)

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:599]: yeah best we can do is some sort of manual cleanup async call and a Drop impl that logs when they haven't been cleaned up. probably worth fixing this while we're in here instead of adding the TODO

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:341]: I think `append_core` and `compare_and_append_core` should be methods on Batch! that feels pretty natural API wise and it neatly solves the naming problem

we could avoid the need to clone Machine, etc by giving Batch a lifetime and having it mutably borrow `&mut WriteHandle`. it seems fine to me to only allow a WriteHandle to be in the process of constructing one batch at a time

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:364]: something I was thinking about after you merged the last PR. I suspect there's some way for us to push some form of this loop into the Machine::compare_and_append call, which I think might make the code a little neater

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:171]: I _think_ we should be able to write this method entirely in terms of compare_and_append_core (i.e. append is something that I'm pretty sure could live completely outside of the persist crate because it's implementable in terms of the _public_ compare_and_append). this would have been clunky before this PR (it would have had to write out the blob again on every expected_upper mismatch), but now IIUC it should be totally doable

**Review by danhhz — COMMENTED** [review]
> I didn't add tests yet but I see two options for that:
> 
> Rewrite all existing tests to use the new API, add a sanity-check to make sure the old-style API works as expected
> Leave all tests as they are, add sanity-check tests to make sure the new-style API works as expected
> Option 2. would be less disruptive but feels wrong because the old-style APIs are really sugar on top of the new APIs, so it seems a sanity-check test is enough for them.
> 
> Option 1. is less disruptive in terms of number of lines of code we'd have to change.

I suspect we'll probably want to keep the "sugar" ones around permanently for simple usages. this also means we don't have to change every existing caller, just opt them in when we feel it's worthwhile

tests will probably mostly continue to use `expect_append` and `expect_compare_and_append`, which could internally either use the sugar or the batch stuff directly (not sure I have an opinion on that yet). this probably eliminates a large part of the distinction between 1 and 2

File: src/persist-client/src/write.rs
Comment (aljoscha) [reply src/persist-client/src/write.rs:419]: This was quite clear! 👍

File: src/persist-client/src/write.rs
Comment (aljoscha) [reply src/persist-client/src/write.rs:336]: I moved these methods to `Batch`, but requiring `Batch` to have a reference to `WriteHandle` felt a bit unnatural, mostly because a successful append changes the cached upper of the `WriteHandle`, so `Batch` has to "reach" into the `WriteHandle` to do that.

I played around with the `concurrency()` test, and that seemed easier with `Batch` not referencing a write handle, though that could be a function of how I implemented it. The last commit moves the `_core` methods back to `WriteHandle` and changes the test, so you can look at it in isolation. At this point it's probably easier to figure out the fine details in a call, but I think I have a good understanding of the ergonomics now.

**Comment by aljoscha** [comment]
@danhhz I pushed a version that has everything but I'm not really happy with how dropping of batches/consuming them works. I'm done for the day, but maybe you'll have some bright idea.

**Comment by aljoscha** [comment]
No worries, this gave me some time to clean this up a bit more. 😊

File: src/persist-client/examples/maelstrom/txn.rs
Comment (aljoscha) [inline src/persist-client/examples/maelstrom/txn.rs:152]: I don't like that this change is necessary, even though the public API didn't change. (It's bleeding through in the Future type of the function)

When this is not changed, we get:
```
error: implementation of `FnOnce` is not general enough
   --> src/persist-client/examples/maelstrom/txn.rs:448:66
    |
448 |       async fn eval(&self, handle: Handle, src: NodeId, req: Body) {
    |  __________________________________________________________________^
449 | |         match req {
450 | |             Body::ReqTxn { msg_id, txn } => {
451 | |                 let in_reply_to = msg_id;
...   |
469 | |         }
470 | |     }
    | |_____^ implementation of `FnOnce` is not general enough
    |
    = note: closure with signature `fn(&'0 (MaelstromKey, MaelstromVal, i64)) -> ((&MaelstromKey, &MaelstromVal), &u64, &i64)` must implement `FnOnce<(&(MaelstromKey, MaelstromVal, i64),)>`, for any lifetime `'0`...
    = note: ...but it actually implements `FnOnce<(&(MaelstromKey, MaelstromVal, i64),)>`
```

And I haven't yet figured out why that is. It must have something to do with how the internal code structure behind the function call is different now.

File: src/persist-client/src/bin/persist_open_loop_benchmark.rs
Comment (danhhz) [inline src/persist-client/src/bin/persist_open_loop_benchmark.rs:454]: should something be joining on these during shutdown? (probably)

File: src/persist-client/src/lib.rs
Comment (danhhz) [inline src/persist-client/src/lib.rs:947]: nit: you've got some debugging here

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:135]: let's add something like a `batch_from_iter` (maybe even could just be named `batch`?) method to WriteHandle that both of these can use

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:452]: nit: empty line between these

I'm down with reserving async for the future, but curious what you're imagining here? in my mind, `BatchBuilder::add` being async should be sufficient, but I very well could be missing something

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:519]: I think this `pub(crate)` can be removed?

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:597]: I think the `pub(crate)` can be removed from all these fields?

**Review by danhhz — COMMENTED** [review]
I might have one more round of nits in me, but feels like this is getting really close!

> I'm not really happy with how dropping of batches/consuming them works. I'm done for the day, but maybe you'll have some bright idea.

what you have is basically exactly what I was imagining!

File: src/persist-client/src/lib.rs
Comment (aljoscha) [reply src/persist-client/src/lib.rs:365]: 🙈 good catch! this was me trying something that I expected to work, shouldn't be in there.

File: src/persist-client/src/write.rs
Comment (aljoscha) [reply src/persist-client/src/write.rs:383]: > as for naming, let's call this compare_and_append_batch (ditto append_batch)

will do!

File: src/persist-client/src/write.rs
Comment (aljoscha) [reply src/persist-client/src/write.rs:404]: > something I was thinking about after you merged the last PR. I suspect there's some way for us to push some form of this loop into the Machine::compare_and_append call, which I think might make the code a little neater


done!

File: src/persist-client/src/write.rs
Comment (aljoscha) [reply src/persist-client/src/write.rs]: > I think we should be able to write this method entirely in terms of compare_and_append_core (i.e. append is something that I'm pretty sure could live completely outside of the persist crate because it's implementable in terms of the public compare_and_append). this would have been clunky before this PR (it would have had to write out the blob again on every expected_upper mismatch), but now IIUC it should be totally doable

done

File: src/persist-client/src/write.rs
Comment (aljoscha) [reply src/persist-client/src/write.rs:574]: > either it should take a writehandle or the batch should keep a copy internally of the Arc. probably the latter

changed to keep a blob Arc in the batch

File: src/persist-client/examples/maelstrom/txn.rs
Comment (aljoscha) [reply src/persist-client/examples/maelstrom/txn.rs:152]: I'm quite sure it's because the borrowed references are now held across await points (because all of `create_builder`, `add`, and `finish` are `async`), which makes the lifetime bleed into the Future signature of `compare_and_append`. Before, they were borrowed in `encode_batch()`, but never across an await point.

File: src/persist-client/src/write.rs
Comment (aljoscha) [reply src/persist-client/src/write.rs:716]: > unless I'm missing something, this shouldn't need all the Borrow params. just take a k: &K, v: &V, t: &T, d: &D and make the caller deal with it

changed!

Comment (aljoscha) [reply src/persist-client/src/write.rs:716]: I think `AsRef<>` or `Borrow<>` would allow more flexibility for the caller, but I don't have a strong opinion so will go with simple references for now.

I still have this rust beginner -> rust pro thing in my head where you go:
 - `String`
 - `&str`
 - `AsRef<str>`
 - maybe then `Cow<'a, str>`

😅

File: src/persist-client/src/write.rs
Comment (aljoscha) [reply src/persist-client/src/write.rs:597]: I removed all `pub(crate)`/`pub` except `pub desc` on `Batch`. Does that sound alright?

File: src/persist-client/examples/maelstrom/txn.rs
Comment (danhhz) [reply src/persist-client/examples/maelstrom/txn.rs:152]: ah okay. well then this probably would have happened when we did the "bounded memory usage" stuff because that would also introduce an await point (when the columnar batch fills up and is flushed out into a blob)

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:376]: they seem fine to me!

File: src/persist-client/src/write.rs
Comment (danhhz) [reply src/persist-client/src/write.rs:409]: might be a good followup for you? it basically amounts to the sort of `builder.take_filled()` logic we have in old persist trace compaction. definitely M2 but probably doesn't need to be M1

https://github.com/MaterializeInc/materialize/blob/b34b247a9249200fea77145e0c13eb6014e88019/src/persist/src/indexed/background.rs#L307-L320

File: src/persist-client/src/bin/persist_open_loop_benchmark.rs
Comment (danhhz) [reply src/persist-client/src/bin/persist_open_loop_benchmark.rs:469]: can this be dead_code be removed now?

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:445]: did we lose the actual +1 at some point?

File: src/persist-client/src/write.rs
Comment (danhhz) [reply src/persist-client/src/write.rs]: `desc` is still pub because the persist user might want to be able to see the bounds? It's a little scary that they could muck around with it (plus the since is sorta meaningless). I'd say move that one non-pub, too, and add upper and lower `pub fn`s to Batch

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:260]: this is definitely true for `compare_and_append_batch` but is it also true for `append_batch`? I think maybe not but I don't want to be prescriptive about it yet. keep this as-is, but I wanted to get you thinking about it

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:577]: should this take `self`? I can't see why someone would want to reuse the batch after, can you? (the builder is eventually a different story, lots of nice allocs there to be potentially reused)

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:725]: oh this might be my fault. I think we still need some sort of `//` or `///` prefix here before the empty line for rustdoc to work :)

File: src/persist-client/src/write.rs
Comment (danhhz) [inline src/persist-client/src/write.rs:729]: nit: remove maybe?

**Review by danhhz — APPROVED** [review]
this ended up being more work than I was initially imagining but it's ended up in a really good place!

LGTM modulo a few last things. the ones I care about most are the couple panics in response to user input that should be InvalidUsage and (to a lesser extent) the pub desc

File: src/persist-client/src/bin/persist_open_loop_benchmark.rs
Comment (aljoscha) [reply src/persist-client/src/bin/persist_open_loop_benchmark.rs:407]: We can, but it's a tiny but more involved because writers are currently kept in a `Vec<Box<dyn BenchmarkWriter ...>>`. Will change, though.

File: src/persist-client/src/write.rs
Comment (aljoscha) [reply src/persist-client/src/write.rs:577]: The `_append()` methods take a `&mut Batch`, mostly so that callers can keep the batch and retry. And the cleanup-check logic becomes tricky if `Batch` is cloned. But I don't like that `delete` takes a `&mut self`, either. :see_no_evil:

File: src/persist-client/src/write.rs
Comment (danhhz) [reply src/persist-client/src/write.rs:260]: I agree the contract is simple, but making the caller deal with the batch is not simple for them. If the contract of this method can be "we always use/delete the batch" that's simpler for them :)

Comment (aljoscha) [reply src/persist-client/src/write.rs:260]: True! :sweat_smile: 

I could change `append_batch` to take a `self` and keep `compare_and_append_batch` on `&mut self` to make that distinction clear. Our internal logic would be slightly more complicated in `append_batch` but that should be worth it.

File: src/persist-client/src/write.rs
Comment (aljoscha) [reply src/persist-client/src/write.rs]: With this change: https://github.com/MaterializeInc/materialize/pull/12477#discussion_r879545137. It seems I can make `delete` take `self`, because all places that want to call it own the `Batch`.

**Pull request merged**
